File: src/utils/generate_prompt.py
from config.platform_syntax import platform_syntax

def get_platform_syntax(platform):
    """
    Retrieves the syntax for the specified platform.

    Args:
        platform (str): The platform name (MUST BE ONE OF "Slack", "Discord").

    Returns:
        str: A prompt containing the syntax for the specified platform.
    """
    return (
        f"Syntax for {platform}:\n"
        f"Links: {platform_syntax[platform].get('links', "Unsupported on this platform.")}\n"
        f"Bold: {platform_syntax[platform].get('bold', 'Unsupported on this platform.')}\n"
        f"Italic: {platform_syntax[platform].get('italic', 'Unsupported on this platform.')}\n"
        f"Code: {platform_syntax[platform].get('code', 'Unsupported on this platform.')}\n"
        f"Code Block: {platform_syntax[platform].get('code_block', 'Unsupported on this platform.')}\n"
        f"List: {platform_syntax[platform].get('list', 'Unsupported on this platform.')}\n"
        f"Numbered List: {platform_syntax[platform].get('numbered_list', 'Unsupported on this platform.')}\n"
        f"Quote: {platform_syntax[platform].get('quote', 'Unsupported on this platform.')}\n"
        f"Strikethrough: {platform_syntax[platform].get('strikethrough', 'Unsupported on this platform.')}\n"
        f"Underline: {platform_syntax[platform].get('underline', 'Unsupported on this platform.')}\n"
        f"Spoiler: {platform_syntax[platform].get('spoiler', 'Unsupported on this platform.')}\n"
        f"Header: {platform_syntax[platform].get('header', 'Unsupported on this platform.')}\n"
        f"Remember to use the appropriate formatting for {platform} so that your message renders correctly for the user."
    )

def generate_prompt(username, platform):
    """
    Generates a system message for the AI model based on the platform and username.

    Args:
        username (str): The username of the user.
        platform (str): The platform name (MUST BE ONE OF "Slack", "Discord").

    Returns:
        str: The generated system message.
    """
    return f"""You are **Iris**, a helpful assistant bot operating on the {platform} platform. You are an AI companion acting as part of the Deepgram team.

You assist users like {username} by guiding them to solve problems through self-discovery, not just direct answers.

**Behavioral Guidelines**:
✅ Always:
• Use the user's name ({username}) in every reply.
• Default to `curl` for HTTP requests and Python for WebSocket examples, unless another language is requested.
• Provide valid and accessible links to any sources you mention.
• Format output using the correct {platform} syntax.
• Keep responses under 2000 characters.
• Expand with detailed answers only after the user asks follow-up questions.

⚠️ Rarely:
• Offer direct answers—prompt the user to think or try first.
• Assume the user's technical skill—let them show or tell you.

⛔️ Never:
• Use unsupported formatting or features for {platform}.
• Include broken or inaccessible links.
• Ignore the formatting syntax guide below.

{get_platform_syntax(platform)}

Each message will contain front-matter with the following fields:
- user: The user who sent the message.
- date: The timestamp of the message.
- channel: The channel where the message was sent.
- mentions: Whether the user mentioned you specifically, thereby requesting a response from you.
"""

File: src/utils/generatePrompt.ts
/**
 * @copyright Deepgram
 * @license MIT
 * @author Naomi Carrigan
 */

import { platformSyntax } from "../config/platformSyntax.js";
import { isPropertyInPlatformSyntaxObject } from "./typeguards.js";

type Platform = "discord" | "slack";

/**
 * Gets a platform property value or returns a short unsupported marker.
 * @param platform - The platform to check (either "discord" or "slack").
 * @param property - The property to look up.
 * @returns The value of the property for the platform, or "—" if unsupported.
 */
const lookup = (platform: Platform, property: string): string => {
  return isPropertyInPlatformSyntaxObject(platform, property)
    ? platformSyntax[platform][property]
    : "—";
  // Em dash for unsupported
};

/**
 * Generates a compact syntax reference block for the platform.
 * @param platform - The platform to generate the syntax block for.
 * @returns A string containing the syntax block.
 */
const syntaxBlock = (platform: Platform): string => {
  return `${platform.toUpperCase()} message formatting:
block: ${lookup(platform, "block")}
bold: ${lookup(platform, "bold")}
code: ${lookup(platform, "code")}
header: ${lookup(platform, "header")}
italic: ${lookup(platform, "italic")}
link: ${lookup(platform, "link")}
list: ${lookup(platform, "list")}
numberedList: ${lookup(platform, "numberedList")}
quote: ${lookup(platform, "quote")}
spoiler: ${lookup(platform, "spoiler")}
strikethrough: ${lookup(platform, "strikethrough")}
underline: ${lookup(platform, "underline")}`;
};

/**
 * Generates a prompt for Iris.
 * @param username - The username of the user.
 * @param platform - The platform the user is on (either "discord" or "slack").
 * @returns The generated prompt.
 */
const generatePrompt = (
  username: string,
  platform: Platform,
): string => {
  return `You are Iris on ${platform}. Address ${username} directly.

${syntaxBlock(platform)}

Rules:
• Terse responses ≤2000 chars
• Infer code language; default curl/websocat
• Only provide existing doc links
• Reference prior accurate answers if they exist in conversation
• Ask clarification if unclear

Frontmatter is context — never echo it`;
};

export { generatePrompt };
